File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>>adj(numCourses);
        for(int i=0;i<prerequisites.size();i++){
            int course=prerequisites[i][0];
            int depends=prerequisites[i][1];
            adj[course].push_back(depends);
        }
        vector<int>topo;
        queue<int>q;
        vector<int>indegree(numCourses,0);
        for(int i=0;i<prerequisites.size();i++){
            int in=prerequisites[i][1];
            indegree[in]++;
        }
        for(int i=0;i<indegree.size();i++){
            if(indegree[i]==0){
                q.push(i);
            }
        }
        while(!q.empty()){
            int node=q.front();
            q.pop();
            topo.push_back(node);

            for(auto it:adj[node]){
                indegree[it]--;
                if(indegree[it]==0){
                    q.push(it);
                }
            }
        }
        if(topo.size()<numCourses){
            return false;
        }
        return true;
    }
};

File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        n=numCourses
        adjList=[[] for _ in range(n)]
        for course in prerequisites:
            adjList[course[0]].append(course[1])
        topo=[]
        indegree=[0]*n
        q=deque()
        for course in adjList:
            for adj in course:
                indegree[adj]+=1
        
        for i in range(n):
            if indegree[i]==0:
                q.append(i)
        while q:
            course=q.popleft()
            topo.append(course)
            for adj in adjList[course]:
                indegree[adj]-=1
                if(indegree[adj]==0):
                    q.append(adj)

        return len(topo)==n
            